x/tokenomics/keeper/token_logic_modules.go

```diff
@@ -319,149 +337,216 @@ func (k Keeper) TokenLogicModuleRelayBurnEqualsMint(
 
-	accountAddr, err := cosmostypes.AccAddressFromBech32(addr)
+	accountAddr, err := cosmostypes.AccAddressFromBech32(destAdr)
 	if err != nil {
-		return nil, err
+		return sdk.Coin{}, err
 	}
 
-	coinsToAccAmt := calculateGlobalMintAllocationFromSettlementAmount(settlementAmtFloat, allocation)
+	coinsToAccAmt := calculateAllocationAmount(settlementAmtFloat, allocation)
 	coinToAcc := cosmostypes.NewCoin(volatile.DenomuPOKT, math.NewInt(coinsToAccAmt))
 	if err := k.bankKeeper.SendCoinsFromModuleToAccount(
-		ctx, suppliertypes.ModuleName, accountAddr, sdk.NewCoins(coinToAcc),
+		ctx, srcModule, accountAddr, sdk.NewCoins(coinToAcc),
 	); err != nil {
-		return nil, err
+		return sdk.Coin{}, err
 	}
-	logger.Info(fmt.Sprintf("sent (%v) coins from the tokenomics module to the account with address %q", coinToAcc, addr))
+	logger.Info(fmt.Sprintf("sent (%v) coins from the tokenomics module to the account with address %q", coinToAcc, destAdr))
 
-	return &coinToAcc, nil
+	return coinToAcc, nil
 }
 
-func (k Keeper) handleOverservicedApplication(
+// ensureClaimAmountLimits checks if the application was overserviced and handles
+// the case if it was.
+// Per Algorithm #1 in the Relay Mining paper, the maximum amount that a single supplier
+// can claim in a session is AppStake/NumSuppliersPerSession.
+// If this is not the case, then the supplier essentially did "free work" and the
+// actual claim amount is less than what was claimed.
+// Ref: https://arxiv.org/pdf/2305.10672
+func (k Keeper) ensureClaimAmountLimits(
 	ctx context.Context,
+	logger log.Logger,
 	application *apptypes.Application,
-	settlementCoins cosmostypes.Coin,
+	supplier *sharedtypes.Supplier,
+	claimSettlementCoin cosmostypes.Coin,
 ) (
-	newSettlementCoins cosmostypes.Coin,
+	actualSettlementCoins cosmostypes.Coin,
 	err error,
 ) {
-	logger := k.Logger().With("method", "handleOverservicedApplication")
-	// over-serviced application
-	logger.Warn(fmt.Sprintf(
-		"THIS SHOULD NEVER HAPPEN. Application with address %s needs to be charged more than it has staked: %v > %v",
-		application.Address,
-		settlementCoins,
-		application.Stake,
-	))
-
-	// TODO_MAINNET(@Olshansk, @RawthiL): The application was over-serviced in the last session so it basically
-	// goes "into debt". Need to design a way to handle this when we implement
-	// probabilistic proofs and add all the parameter logic. Do we touch the application balance?
-	// Do we just let it go into debt? Do we penalize the application? Do we unstake it? Etc...
-	// See this document from @red-0ne and @bryanchriswhite for more context: notion.so/buildwithgrove/Off-chain-Application-Stake-Tracking-6a8bebb107db4f7f9dc62cbe7ba555f7
-	expectedBurn := settlementCoins
+	logger = logger.With("helper", "ensureClaimAmountLimits")
+
+	// TODO_BETA_OR_MAINNET(@red-0ne): The application stake gets reduced with every claim
+	// settlement. Relay miners use the appStake at the beginning of a session to determine
```

Comment:
TL;DR, I think we need to consider adding support for in-session stake increases (i.e. taking stake increases into account within the session they're committed).

I find myself thinking about this from the perspective of the application actor's UX, which seems to factor into how we handle overserviced applications (e.g. unstake them, etc.) and minimum stake balances:

I think this fact (`Relay miners use...`) means that an application operator who would top up their stake from an insufficient balance to a sufficient one in session N will have to wait for session N+1 before they can expect to be serviced by rational/pessimistic/non-altruistic suppliers?

It's been a minute since I looked at the session hydration code but I also find myself wondering about whether and how app stake balance is/should factor into session generation. It could be the case that an application is able to "get a session" via query but that suppliers may not service them until the next session.

Reply:
```diff
type Session struct {
	Header              *SessionHeader      `protobuf:"bytes,1,opt,name=header,proto3" json:"header,omitempty"`
	SessionId           string              `protobuf:"bytes,2,opt,name=session_id,json=sessionId,proto3" json:"session_id,omitempty"`
	SessionNumber       int64               `protobuf:"varint,3,opt,name=session_number,json=sessionNumber,proto3" json:"session_number,omitempty"`
	NumBlocksPerSession int64               `protobuf:"varint,4,opt,name=num_blocks_per_session,json=numBlocksPerSession,proto3" json:"num_blocks_per_session,omitempty"`
	Application         *types1.Application `protobuf:"bytes,5,opt,name=application,proto3" json:"application,omitempty"`
	Suppliers           []*types.Supplier   `protobuf:"bytes,6,rep,name=suppliers,proto3" json:"suppliers,omitempty"`
+	// TODO_MAINNET(@bryanchriswhite): Add a RelayMining field that can be used for UX/DevX
+	// purposes capturing things like:
+	// - The probability of a proof being required for this session
+	// - The threshold of the probability that triggers a proof definitively
+	// - The maximum claimable pokt per supplier in this session
+	// See the discussion here: https://github.com/pokt-network/poktroll/pull/755#discussion_r1743286043
}
```